gnc/sub8_perception/nodes/marker_occ_grid.py

```diff
@@ -0,0 +1,160 @@
+        center = np.array([pose_2d.x, -pose_2d.y])  # The negative all depends on how the center is returned
+        rotation = pose_2d.theta
+
+        center_offset = center / self.meta_data.resolution - np.array([self.mid_x, self.mid_y])
+
+        rot_top_point = np.dot(np.array([TRENCH_LENGTH, 0]) / 2, make_2D_rotation(rotation)).astype(np.int8)
+        rot_bottom_point = np.dot(-np.array([TRENCH_LENGTH, 0]) / 2, make_2D_rotation(rotation)).astype(np.int8)
+
+        pos_top_point = np.int0(rot_top_point + center_offset)
+        pos_bottom_point = np.int0(rot_bottom_point + center_offset)
+
+        cv2.line(self.trenches, tuple(pos_top_point), tuple(pos_bottom_point), 101, int(TRENCH_WIDTH))
+
+    def publish_grid(self):
+        '''
+        Take the occupancy grid and send it out over ros with timestamps and whatnot.
+        '''
+        t = rospy.Time.now()
+        header = Header(stamp=t, frame_id='/map')
+        # Populate occ grid msg
+        occ_msg = OccupancyGrid()
+        occ_msg.header = header
+        occ_msg.info = self.meta_data
+        # Make sure values don't go out of range
+        occ_grid = self.occ_grid + self.searched + self.trenches
+        occ_msg.data = np.clip(occ_grid.flatten(), -1, 100)
+        self.occ_grid_pub.publish(occ_msg)
+
+
+class MarkerOccGrid(OccGridUtils):
+    '''
+    Handles updating occupancy grid when new data comes in.
+    TODO: Upon call can return some path to go to in order to find them.
+    '''
+    def __init__(self, res, width, height, starting_pose):
+        super(self.__class__, self).__init__(res=res, width=width, height=height, starting_pose=starting_pose)
+
+        self.tf_listener = tf.TransformListener()
+
+        self.image = np.zeros((100, 100))
+    
+    def update_grid(self, marker):
+        '''
+        Takes marker information to update occupacy grid.
+        '''
+        self.tf_listener.waitForTransform("/map", "/down_cam", rospy.Time(), rospy.Duration(1.0))
+        trans, rot = self.tf_listener.lookupTransform("/map", "/down_cam", rospy.Time())
+        
+        x_y_position = trans[:2]
+        depth = trans[2]
+
+        self.add_circle(x_y_position, self.calculate_visual_radius())
+        self.publish_grid()
+
+    def calculate_visual_radius(self):
+        '''
+        Draws rays to find the radius of the FOV of the camera in meters.
+        
+        TODO: Implement with actual ros images and image info.
+        '''
+        depth = self.depth
+        edge_point = np.array([0, self.image.shape[0] / 2])
+
+        mid_ray = unit_vector(projectPixelTo3dRay(self.image / 2, 0, self.depth))
+        edge_ray = unit_vector(projectPixelTo3dRay(edge_point, 1, self.depth))
+
+        # Calculate angle between vectors and use that to find r
+        theta = np.arccos(np.dot(mid_ray, edge_ray))
+        return np.tan(theta) * depth
+
+    def check_for_trench(self):
+        self.trench_service('trench')
+
+
+def projectPixelTo3dRay(uv, temp, depth):
```

Comment:
This was just to return different vectors for each call of the function since the image geometry class was not yet used. When the real image and camera_info exist in the script, this temp function will be removed and the function calls will be modified.